src/main/java/io/gingersnapproject/service/MembershipListener.java

```diff
@@ -0,0 +1,131 @@
+import java.util.concurrent.TimeUnit;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+
+import io.gingersnapproject.cdc.ReschedulingTask;
+import io.gingersnapproject.cdc.configuration.Configuration;
+
+import io.quarkus.runtime.ShutdownEvent;
+import io.quarkus.runtime.StartupEvent;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
+public class MembershipListener {
+
+   private static final Logger log = LoggerFactory.getLogger(MembershipListener.class);
+   private static final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(r -> new Thread(r,"membership-listener"));
+
+   // Accessed by a single thread.
+   private final Set<InetAddress> members = new HashSet<>();
+   @Inject Configuration configuration;
+
+   // TODO: remove instance wrapper once implementation available.
+   @Inject Instance<DynamicMembershipManagement> dmm;
+
+   @ConfigProperty(name = "gingersnap.dynamic-membership", defaultValue = "false")
+   boolean dynamicMembership;
+
+   private ReschedulingTask<Void> task;
+   private volatile boolean shutdown;
+
+   void onStartup(@Observes StartupEvent ignore) {
+      if (task != null || dmm.isUnsatisfied()) return;
+
+      if (dynamicMembership) {
+         task = new ReschedulingTask<>(executor, this::pollDNSInformation, __ -> false, 15, TimeUnit.SECONDS, __ -> true);
+         executor.submit(() -> {
+            task.schedule();
+         });
+      } else {
+         dmm.get().addMember(configuration.cache().uri());
+      }
+   }
+
+   void onShutdown(@Observes ShutdownEvent ignore) {
+      if (task == null) return;
+
+      log.info("Shutdown membership listener");
+      task.close();
+      task = null;
+      shutdown = true;
+   }
+
+   // package-private for testing only.
+   CompletionStage<Void> pollDNSInformation() {
+      if (shutdown) return CompletableFutures.completedNull();
+
+      var uri = configuration.cache().uri();
+      Set<InetAddress> joiners = new HashSet<>(resolve(uri.getHost()));
+      Set<InetAddress> leavers = new HashSet<>(members);
+
+      leavers.removeIf(joiners::contains);
+      joiners.removeIf(members::contains);
+
+      members.removeAll(leavers);
+      members.addAll(joiners);
+
+      log.debug("Updating membership, joined '{}' and left '{}', current '{}'", joiners, leavers, members);
+      var manager = dmm.get();
+      for (InetAddress leaver : leavers) {
+         manager.removeMember(asURI(uri, leaver));
+      }
+
+      for (InetAddress joiner : joiners) {
+         manager.addMember(asURI(uri, joiner));
+      }
+
+      return CompletableFutures.completedNull();
+   }
+
+   public static List<InetAddress> resolve(String query) {
```

Comment:
Why is this static? It seems to be complicating the test code a lot. Do we expect someone to call this method outside of this class without the bean? I would expect to just inject the MembershipListener bean if I needed to use it.

Reply:
We can leave it as a class method. I'll make it package-private, or we'll need to mock the `InetAddress.getAllByName` call.

Reply:
Sure, np